docs/advanced/UsageWithTypescript.md

```diff
@@ -0,0 +1,206 @@
+
+The full source code is available on [codesandbox here](https://codesandbox.io/s/w02m7jm3q7). Note that by going through this example yourself you will experience some of the benefits of using TypeScript.
+
+## Type Checking State
+
+Type checking each slice of state is simple when you understand that each slice of state is just an object.
+
+Describing the state shape of the chat reducer's slice of state:
+
+```ts
+// src/store/chat/types.ts
+
+export interface Message {
+  user: string
+  message: string
+  timestamp: number
+}
+
+export interface ChatState {
+  messages: Message[]
+}
+```
+
+Describing the state shape of the system reducer's slice of state:
+
+```ts
+// src/store/system/types.ts
+
+export interface SystemState {
+  loggedIn: boolean
+  session: string
+  userName: string
+}
+```
+
+Note that we are exporting these interfaces to reuse them later in reducers and action creators.
+
+## Type Checking Actions & Action Creators
+
+We will be using TypeScript's enums to declare our action constants. [Enums](https://www.typescriptlang.org/docs/handbook/enums.html) allow us to define a set of named constants.
+
+Chat Action Constants & Shape:
+
+```ts
+// src/store/chat/types.ts
+
+export enum ChatActions {
+  SendMessage = 'SEND_MESSAGE'
+}
+
+interface SendMessageAction {
+  type: ChatActions.SendMessage
+  payload: Message
+}
+
+export type ChatActionTypes = SendMessageAction
+```
+
+Note that you can use TypeScript's Union Type to express multiple actions like so:
+
+```ts
+export type ChatActionTypes = SendMessageAction | DeleteMessageAction
+```
+
+With these types declared we can now also type check chat's action creators. In this case we are taking advantage of TypeScript's inference to avoid verbosity:
+
+```ts
+// src/store/chat/actions.ts
+
+import { Message, ChatActions } from './types'
+
+export function sendMessage(newMessage: Message) {
+  return {
+    type: ChatActions.SendMessage,
+    payload: newMessage
+  }
+}
+```
+
+System Action Constants & Shape:
+
+```ts
+// src/store/system/types.ts
+
+export enum SystemActions {
+  UpdateSession = 'UPDATE_SESSION'
+}
+
+interface UpdateSessionAction {
+  type: SystemActions.UpdateSession
+  payload: SystemState
+}
+
+export type SystemActionTypes = UpdateSessionAction
+```
+
+With these types we can now also type check system's action creators:
+
+```ts
+// src/store/system/actions.ts
+
+import { SystemActions, SystemState } from './types'
+
+export function updateSession(newSession: SystemState) {
+  return {
+    type: SystemActions.UpdateSession,
+    payload: newSession
+  }
+}
+```
+
+## Type Checking Reducers
+
+Reducers are just pure functions that take the previous state, an action and then return the next state.
+
+Type checked chat reducer:
```

Comment:
Add some description here explaining what exactly is being done in comparison to a plain JS reducer (ie, importing the state and action types and referencing them).